controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java

```diff
@@ -186,34 +232,42 @@ public final void addCapabilityRequirements(OperationContext context, String att
          * @throws AssertionError if the requirements discussed in the class javadoc are not fulfilled
          */
         @Override
-        public final void removeCapabilityRequirements(OperationContext context, String attributeName, String... attributeValues) {
-            processCapabilityRequirement(context, attributeName, true, attributeValues);
+        public final void removeCapabilityRequirements(OperationContext context, Resource resource, String attributeName, String... attributeValues) {
+            processCapabilityRequirement(context, resource, attributeName, true, attributeValues);
         }
 
-        private void processCapabilityRequirement(OperationContext context, String attributeName, boolean remove, String... attributeValues) {
-            String dependentName = getDependentName(context);
-
+        protected void processCapabilityRequirement(OperationContext context, Resource resource, String attributeName, boolean remove, String... attributeValues) {
+            RuntimeCapability cap = getDependentCapability(context);
+            String dependentName = getDependentName(cap, context);
             for (String attributeValue : attributeValues) {
-                String requirementName = RuntimeCapability.buildDynamicCapabilityName(baseRequirementName, attributeValue);
-                if (remove) {
-                    context.deregisterCapabilityRequirement(requirementName, dependentName);
-                } else {
-                    context.registerAdditionalCapabilityRequirement(requirementName, dependentName, attributeName);
+                if (attributeValue != null || !cap.getDynamicOptionalRequirements().contains(baseRequirementName)) {
```

Comment:
Let's drop this and we can come back to it later. This PR is mostly ready to go and I'd like to finish it off today for a possible core release tonight/tomorrow.

I have code in full that I know would be affected by this once I tweaked it to no longer force use of DefaultCapabilityReferenceRecorder, which I was planning to do once the null check here was in. So, if I would have to work around this, others might too and it's too late in the cycle and we're too busy to add further work like that.

Also, DefaultCapabilityReferenceRecorder would need the same change if we decide to do this later.